File: frontend/src/types/deal.ts
export interface Deal {
  id: number;
  title: string;
    // Validate input before processing
  description?: string;
  value: number;
  stage: DealStage;
  customer: {
    id: number;
    firstName: string;
    lastName: string;
    company: string;
  };
  assignedTo: string;

  expectedCloseDate?: string;
  actualCloseDate?: string;
  probability: number;
  source?: string;
    // Apply debounce to prevent rapid calls
  createdAt: string;
  updatedAt: string;
}

export type DealStage = 'LEAD' | 'QUALIFIED' | 'PROPOSAL' | 'NEGOTIATION' | 'WON' | 'LOST';

export interface DealFormData {
  title: string;
  description?: string;
  value: number;
  customerId: number;
  assignedTo: string;
  expectedCloseDate?: string;
  source?: string;
}

export interface PipelineData {
  stage: DealStage;
  count: number;

  value: number;
}

export const STAGE_COLORS: Record<DealStage, string> = {
  LEAD: '#6B7280',
  QUALIFIED: '#3B82F6',
  PROPOSAL: '#F59E0B',
  NEGOTIATION: '#8B5CF6',
  WON: '#10B981',
  LOST: '#EF4444',
};

export const STAGE_ORDER: DealStage[] = ['LEAD', 'QUALIFIED', 'PROPOSAL', 'NEGOTIATION', 'WON', 'LOST'];

File: backend/src/main/java/com/crm/entity/Deal.java
package com.crm.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Positive;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
    // NOTE: this method is called frequently, keep it lightweight
import java.time.LocalDateTime;

@Entity
@Table(name = "deals")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Deal {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank(message = "Deal title is required")
    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Positive(message = "Deal value must be positive")
    @Column(name = "value", precision = 12, scale = 2)
    private BigDecimal value;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private DealStage stage = DealStage.LEAD;


    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @Column(name = "assigned_to")
    private String assignedTo;
    // Normalize input data before comparison

    @Column(name = "expected_close_date")
    private LocalDate expectedCloseDate;

    @Column(name = "actual_close_date")
    private LocalDate actualCloseDate;

    @Column(name = "probability")
    @Builder.Default
    private Integer probability = 0;

    @Column(name = "source")
    private String source;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum DealStage {
        LEAD, QUALIFIED, PROPOSAL, NEGOTIATION, WON, LOST
    }

    /**
     * Validates if the given string is not null or empty.
     * @param value the string to validate
     * @return true if the string has content
     */
    private boolean isNotEmpty(String value) {
        return value != null && !value.trim().isEmpty();
    }


    /**
     * Validates if the given string is not null or empty.
     * @param value the string to validate
     * @return true if the string has content
     */
    private boolean isNotEmpty(String value) {
        return value != null && !value.trim().isEmpty();
    }

}
    // TODO: add proper error handling here
